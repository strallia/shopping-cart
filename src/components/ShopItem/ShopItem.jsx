import { useState } from "react";
import PropTypes from "prop-types";

const ShopItem = ({ item, setItemsData, isForShopPage = false }) => {
  const [addendQuantity, setAddendQuantity] = useState(1);
  const [totalQuantity, setTotalQuantity] = useState(item.quantity);
  const { id, image, title, price, description, quantity } = item;

  const handleAddendQuantityOnChange = (e) => {
    setAddendQuantity(+e.target.value);
  };

  const handleTotalQuantityOnChange = (e) => {
    const updatedTotalQuantity = +e.target.value;
    setTotalQuantity(updatedTotalQuantity);
    setItemsData((items) => {
      return items.map((item) =>
        item.id === id ? { ...item, quantity: updatedTotalQuantity } : item
      );
    });
  };

  const handleAddToCartClick = () => {
    setItemsData((items) => {
      return items.map((item) =>
        item.id === id ? { ...item, quantity: quantity + addendQuantity } : item
      );
    });
    setAddendQuantity(1);
  };

  const handleDeleteItemClick = () => {};

  return (
    <div>
      <img src={image} />
      <p>{title}</p>
      {isForShopPage && <p>description: {description}</p>}
      <label>
        Quantity:
        <input
          id={isForShopPage ? "addend-quantity" : "total-quantity"}
          type="number"
          value={isForShopPage ? addendQuantity : totalQuantity}
          min="1"
          onChange={
            isForShopPage
              ? handleAddendQuantityOnChange
              : handleTotalQuantityOnChange
          }
        />
      </label>
      <p>${price}</p>
      {isForShopPage ? (
        <button onClick={handleAddToCartClick}>Add to Cart</button>
      ) : (
        <button onClick={handleDeleteItemClick}>Delete</button>
      )}
    </div>
  );
};

ShopItem.propTypes = {
  item: PropTypes.shape({
    id: PropTypes.number,
    image: PropTypes.string,
    title: PropTypes.string,
    price: PropTypes.number,
    description: PropTypes.string.isRequired,
    quantity: PropTypes.number,
  }),
  setItemsData: PropTypes.func,
  isForShopPage: PropTypes.bool,
};

export default ShopItem;
