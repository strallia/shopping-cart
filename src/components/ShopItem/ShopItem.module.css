.container {
  background-color: var(--pale-yellow);
  border-radius: 20px;
  padding: 20px;
  border: 3px solid black;
  box-shadow: 4px 4px 0 0 black;
  display: grid;
  gap: 15px;
}

.image {
  height: 150px;
  object-fit: contain;
  width: 100%;
  background-color: white;
  padding: 20px;
  border-radius: 20px;
}

.title {
  text-align: center;
  color: var(--dark-purple);
}

.title,
.price,
.addToCartButton {
  font-weight: bold;
}

.price {
  width: max-content;
  margin-right: 10px;
}

.description {
  font-size: 0.8rem;
}

.priceQuantityContainer {
  display: flex;
  justify-content: space-around;
  align-items: center;
}

.priceQuantityContainer,
.addToCartButton {
  align-self: end;
}

.quantity {
  display: flex;
  gap: 15px;
  align-items: center;
}

.addendQuantity {
  width: 50px;
  padding: 3px 5px;
  border-radius: 5px;
  border: 2px solid black;
}

.addToCartButton {
  height: min-content;
  cursor: pointer;
  justify-self: center;
  width: max-content;
  border-radius: 10px;
  padding: 5px 10px;
  border: 2px solid black;
  box-shadow: 3px 3px 0 0 black;
  transition: all 0.1s ease-out;
  background-color: var(--green);
}

.addToCartButton:hover {
  background-color: var(--orange);
}

.addToCartButton:active {
  transform: translate(3px, 3px);
  box-shadow: none;
}
