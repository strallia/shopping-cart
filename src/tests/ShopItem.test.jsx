import { render, screen } from "@testing-library/react";
import ShopItem from "../components/ShopItem/ShopItem";
import { vi } from "vitest";
import userEvent from "@testing-library/user-event";

describe("ShopItem on ShopPage", () => {
  const props = {
    item: {
      id: 0,
      image: "url",
      title: "title",
      price: 123,
      description: "description",
      quantity: 0,
    },
    setItemsData: vi.fn(),
    isForShopPage: true,
  };
  it("renders an item with title, description, quantity, price, and add to cart button", () => {
    const { container } = render(<ShopItem {...props} />);
    expect(container).toMatchSnapshot();
  });

  it("calls setItemsData when add-to-cart button is clicked", async () => {
    const setItemsData = vi.fn();
    const user = userEvent.setup();
    render(<ShopItem {...props} setItemsData={setItemsData} />);
    const addToCartBtn = screen.getByRole("button", { name: "Add to Cart" });
    await user.click(addToCartBtn);
    expect(setItemsData).toHaveBeenCalled();
  });

  it("clicking add-to-cart button resets quantity value to 1", async () => {
    const user = userEvent.setup();
    render(<ShopItem {...props} />);
    const addToCartBtn = screen.getByRole("button", { name: "Add to Cart" });
    const quantityValue = screen.getByRole("spinbutton", { name: "Quantity:" });
    await user.click(addToCartBtn);
    expect(quantityValue.value).toBe("1");
  });

  it("renders default quantity value of 1 on mount", () => {
    render(<ShopItem {...props} />);
    const quantityValue = screen.getByRole("spinbutton", { name: "Quantity:" });
    expect(quantityValue.value).toBe("1");
  });
});

describe("ShopItem on CartPage", () => {
  it("renders an item with title, quantity, price, and remove button", () => {
    const props = {
      item: {
        id: 0,
        image: "url",
        title: "title",
        price: 123,
        description: "description",
        quantity: 0,
      },
      setItemsData: vi.fn(),
      isForShopPage: false,
    };
    const { container } = render(<ShopItem {...props} />);
    expect(container).toMatchSnapshot();
  });

  it("if added item with a quantity of 2 on ShopPage, initially renders that item's quantity with value of 2 on mount", () => {
    const props = {
      item: {
        id: 0,
        image: "url",
        title: "title",
        price: 123,
        description: "description",
        quantity: 2,
      },
      setItemsData: vi.fn(),
      isForShopPage: false,
    };
    render(<ShopItem {...props} />);
    const quantity = screen.getByRole("spinbutton", { name: "Quantity:" });
    expect(quantity.value).toBe("2");
  });

  it("calls setItemsData when remove button is clicked", async () => {
    const props = {
      item: {
        id: 0,
        image: "url",
        title: "title",
        price: 123,
        description: "description",
        quantity: 0,
      },
      setItemsData: vi.fn(),
      isForShopPage: false,
    };
    const setItemsData = vi.fn();
    const user = userEvent.setup();
    render(<ShopItem {...props} setItemsData={setItemsData} />);
    const removeBtn = screen.getByRole("button", { name: "Remove" });
    await user.click(removeBtn);
    expect(setItemsData).toHaveBeenCalled();
  });
});
