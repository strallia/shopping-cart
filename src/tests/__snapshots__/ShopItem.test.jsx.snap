// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ShopItem on CartPage > renders an item with title, quantity, price, and remove button 1`] = `
<div>
  <div
    class="_container_859760"
  >
    <img
      class="_image_859760"
      src="url"
    />
    <p
      class="_title_859760"
    >
      title
    </p>
    <div
      class="_priceQuantityContainer_859760"
    >
      <p
        class="_price_859760"
      >
        $ 
        123
      </p>
      <label
        class="_quantityLabel_859760"
      >
        Quantity:
        <input
          class="_totalQuantity_859760"
          id="0"
          min="1"
          type="number"
          value="0"
        />
      </label>
    </div>
    <button
      class="_removeButton_859760"
    >
      Remove
    </button>
  </div>
</div>
`;

exports[`ShopItem on ShopPage > renders an item with title, description, quantity, price, and add to cart button 1`] = `
<div>
  <div
    class="_container_859760"
  >
    <img
      class="_image_859760"
      src="url"
    />
    <p
      class="_title_859760"
    >
      title
    </p>
    <p
      class="_description_859760"
    >
      description
    </p>
    <div
      class="_priceQuantityContainer_859760"
    >
      <p
        class="_price_859760"
      >
        $ 
        123
      </p>
      <label
        class="_quantityLabel_859760"
      >
        Quantity:
        <input
          class="_addendQuantity_859760"
          id="0"
          min="1"
          type="number"
          value="1"
        />
      </label>
    </div>
    <button
      class="_addToCartButton_859760"
    >
      Add to Cart
    </button>
  </div>
</div>
`;
